packages/lesswrong/server/authentication/getOrCreateForumUser.ts

```diff
@@ -0,0 +1,119 @@
+import type { Profile } from "passport";
+import type { VerifyCallback } from "passport-oauth2";
+import { captureException } from "@sentry/core";
+import { userFindOneByEmail, usersFindAllByEmail } from "../commonQueries";
+import { createMutator, updateMutator } from "../vulcan-lib";
+import Users from "../../lib/vulcan-users";
+import { promisify } from "util";
+
+export type IdFromProfile<P extends Profile> = (profile: P) => string | number;
+
+export type UserDataFromProfile<P extends Profile> = (profile: P) => Promise<Partial<DbUser>>;
+
+const mergeAccount = async (profilePath: string, user: DbUser, profile: Profile) =>
+  updateMutator({
+    collection: Users,
+    documentId: user._id,
+    set: {[profilePath]: profile} as AnyBecauseTodo,
+    // Normal updates are not supposed to update services
+    validate: false,
+  });
+
+/**
+ * If the user's email has been updated by their OAuth provider, change their
+ * email to match their OAuth provider's given email
+ */
+const syncOAuthUser = async (user: DbUser, profile: Profile): Promise<DbUser> => {
+  if (!profile.emails || !profile.emails.length) {
+    return user
+  }
+  // I'm unable to find documenation of how to interpret the emails object. It's
+  // plausible we should always set the users email to the first one, but it
+  // could be that the ordering doesn't matter, in which case we'd want to avoid
+  // spuriously updating the user's email based on whichever one happened to be
+  // first. But if their email is entirely missing, we should update it to be
+  // one given by their OAuth provider. Probably their OAuth provider will only
+  // ever report one email, in which case this is over-thought.
+  const profileEmails = profile.emails.map(emailObj => emailObj.value);
+  if (user.email && !profileEmails.includes(user.email)) {
+    // Attempt to update the email field on the account to match the OAuth-provided
+    // email. This will fail if the user has both an OAuth and a non-OAuth account
+    // with the same email.
+    const preexistingAccountWithEmail = await userFindOneByEmail(profileEmails[0]);
+    if (!preexistingAccountWithEmail) {
+      const updatedUserResponse = await updateMutator({
+        collection: Users,
+        documentId: user._id,
+        set: {
+          email: profileEmails[0],
+          // Will overwrite other past emails which we don't actually want to support
+          emails: [{address: profileEmails[0], verified: true}],
+        },
+        validate: false,
+      });
+      return updatedUserResponse.data;
+    }
+  }
+  return user;
+}
+
+export const getOrCreateForumUser = async <P extends Profile>(
```

Comment:
This is refactored out of `authenticationMiddlewares.ts` to avoid a circular import